include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,603 @@
+typedef enum {
+	/** No hash algorithm selected. */
+	ODP_COMP_HASH_ALG_NONE,
+
+	/** SHA-1 hash algorithm. */
+	ODP_COMP_HASH_ALG_SHA1,
+
+	/** SHA-2 hash algorithm 256-bit digest length. */
+	ODP_COMP_HASH_ALG_SHA256
+} odp_comp_hash_alg_t;
+
+/**
+ * Compression algorithms
+ *
+ */
+typedef enum {
+	/** No algorithm specified. Added for testing purpose. */
+	ODP_COMP_ALG_NULL,
+
+	/** DEFLATE - RFC1951 */
+	ODP_COMP_ALG_DEFLATE,
+
+	/** ZLIB - RFC1950 */
+	ODP_COMP_ALG_ZLIB,
+
+	/** LZS */
+	ODP_COMP_ALG_LZS
+} odp_comp_alg_t;
+
+/**
+ * Compression operation status codes
+ */
+typedef enum {
+	/** Operation completed successfully*/
+	ODP_COMP_STATUS_SUCCESS,
+
+	/** Operation terminated due to insufficient output buffer */
+	ODP_COMP_STATUS_OUT_OF_SPACE_TERM,
+
+	/** Operation failure */
+	ODP_COMP_STATUS_FAILURE,
+} odp_comp_status_t;
+
+/**
+ * Hash algorithms in a bit field structure
+ */
+typedef union odp_comp_hash_algos_t {
+	/** hash algorithms */
+	struct {
+		/** ODP_COMP_HASH_ALG_NONE */
+		uint32_t none	: 1,
+
+		/** ODP_COMP_HASH_ALG_SHA1 */
+		uint32_t sha1	: 1;
+
+		/** ODP_COMP_HASH_ALG_SHA256 */
+		uint32_t sha256	: 1;
+
+	} bit;
+
+	/** All bits of the bit field structure
+	 *
+	 * This field can be used to set/clear all flags, or bitwise
+	 * operations over the entire structure.
+	 */
+	uint32_t all_bits;
+} odp_comp_hash_algos_t;
+
+/**
+ * Compression algorithms in a bit field structure
+ */
+typedef union odp_comp_algos_t {
+	/** Compression algorithms */
+	struct {
+		/** ODP_COMP_ALG_NULL */
+		uint32_t null       : 1;
+
+		/** ODP_COMP_ALG_DEFLATE */
+		uint32_t deflate    : 1;
+
+		/** ODP_COMP_ALG_ZLIB */
+		uint32_t zlib       : 1;
+
+		/** ODP_COMP_ALG_LZS */
+		uint32_t lzs        : 1;
+	} bit;
+
+	/** All bits of the bit field structure
+	 * This field can be used to set/clear all flags, or bitwise
+	 * operations over the entire structure.
+	 */
+	uint32_t all_bits;
+} odp_comp_algos_t;
+
+/**
+ * Compression Interface Capabilities
+ */
+typedef struct odp_comp_capability_t {
+	/** Maximum number of  sessions */
+	uint32_t max_sessions;
+
+	/** Supported compression algorithms */
+	odp_comp_algos_t comp_algos;
+
+	/** Supported hash algorithms */
+	odp_comp_hash_algos_t hash_algos;
+
+	/** Synchronous compression mode support (ODP_COMP_OP_MODE_SYNC) */
+	odp_support_t sync;
+
+	/** Aynchronous compression mode support (ODP_COMP_OP_MODE_SSYNC) */
+	odp_support_t async;
+} odp_comp_capability_t;
+
+/**
+ * Hash algorithm capabilities
+ */
+typedef struct odp_comp_hash_alg_capability_t {
+	/** Digest length in bytes */
+	uint32_t digest_len;
+} odp_comp_hash_alg_capability_t;
+
+/**
+ * Compression dictionary.
+ * The dictionary is a sequence of bytes which are initially fed to the
+ * compressor engine. A preset dictionary is specially useful to compress
+ * short input sequences. The compression engine can take advantage of the
+ * dictionary context to encode the input in a more compact manner. (RFC 1950)
+ */
```

Comment:
It would be useful to add a bit more description from RFC 1950
* The compressor and the decompressor must use exactly the same dictionary.
* The dictionary may be fixed or may be chosen among a certain number of predefined dictionaries, according to the kind of input data.
* the optimal dictionaries are application specific.

Also still some specification of dictionary data content/format is needed.

See for example http://www.zlib.net/manual.html 
" The dictionary should consist of strings (byte sequences) that are likely to be encountered later in the data to be compressed, with the most commonly used strings preferably put towards the end of the dictionary. Using a dictionary is most useful when the data to be compressed is short and can be predicted with good accuracy; the data can then be compressed better than with the default empty dictionary"

How are the strings separated? With '0' characters, etc formatting specification. How long/short strings? Which way strings should be ordered? ...

Reply:
The dictionary support is exported as a capability field. Hence implementation can specify if they can support dictionary or not